gli/io.py

```diff
@@ -0,0 +1,239 @@
+        data[f"Edge_{e.name}"] = e.data
+    if graph_node_lists is None:
+        if len(node_attrs) > 0:
+            num_nodes = node_attrs[0].data.shape[0]
+        else:
+            num_nodes = edge.max() + 1
+            # Warning: the number of nodes is not necessarily equal to the
+            # maximum node ID in `edge` + 1 when there are isolated nodes.
+            warnings.warn("There is no node attributes. The number of nodes "
+                          "is inferred from the maximum node ID in `edge` plus"
+                          " 1. This may not be correct if there are isolated "
+                          "nodes.")
+        data["Graph_NodeList"] = np.ones((1, num_nodes), dtype=np.int32)
+    else:
+        data["Graph_NodeList"] = graph_node_lists
+    if graph_edge_lists is not None:  # _EdgeList is optional in metadata.json.
+        data["Graph_EdgeList"] = graph_edge_lists
+    for g in graph_attrs:
+        assert g.name not in ("NodeList", "EdgeList"), \
+            "The name of a graph attribute cannot be 'NodeList' or 'EdgeList'."
+        data[f"Graph_{g.name}"] = g.data
+    # Call save_data().
+    key_to_loc = save_data(f"{name}__graph", **data)
+
+    def _attr_to_metadata_dict(prefix, a):
+        """Obtain the metadata dict of the attribute.
+
+        Args:
+            prefix (str): The prefix of the attribute, i.e., "Node", "Edge", or
+                "Graph".
+            a (Attribute): The attribute.
+        Returns:
+            dict: The metadata dict of the attribute.
+        """
+        metadata = {}
+        metadata.update(a.get_metadata_dict())
+        metadata.update(key_to_loc[f"{prefix}_{a.name}"])
+        return metadata
+
+    # Create the metadata dict.
+    metadata = {"description": description, "data": {}}
+    # Add the metadata of the node attributes.
+    node_dict = {}
+    for n in node_attrs:
+        node_dict[n.name] = _attr_to_metadata_dict("Node", n)
+    metadata["data"]["Node"] = node_dict
+    # Add the metadata of the edge attributes.
+    edge_dict = {"_Edge": key_to_loc["Edge_Edge"]}
+    for e in edge_attrs:
+        edge_dict[e.name] = _attr_to_metadata_dict("Edge", e)
+    metadata["data"]["Edge"] = edge_dict
+    # Add the metadata of the graph attributes.
+    graph_dict = {"_NodeList": key_to_loc["Graph_NodeList"]}
+    if graph_edge_lists is not None:
+        graph_dict["_EdgeList"] = key_to_loc["Graph_EdgeList"]
+    for g in graph_attrs:
+        graph_dict[g.name] = _attr_to_metadata_dict("Graph", g)
+    metadata["data"]["Graph"] = graph_dict
+
+    metadata["citation"] = citation
+    metadata["is_heterogeneous"] = is_heterogeneous
+
+    if citation == "":
+        warnings.warn("The citation is empty.")
+    if is_heterogeneous:
+        raise NotImplementedError(
+            "Heterogeneous graphs are not supported yet.")
+
+    with open("metadata.json", "w", encoding="utf-8") as f:
+        json.dump(metadata, f, indent=4)
```

Comment:
Should we use an optional `save_dir` keyword to specify the saving path of `metadata.json`?

Reply:
Sounds good. Will add it.

Reply:
Actually, maybe not. The `gli.utils.save_data` does not have an dir argument. Saving npz files and `metadata.json` to different directories might be confusing?

Or maybe I can add save_dir to both save_data and here?

Reply:
Added save_dir to both in [3abd4ca](https://github.com/Graph-Learning-Benchmarks/gli/pull/431/commits/3abd4caf08a4a61311208d6eb49c45496040884a)